lib/llm/src/key_value_store.rs

```diff
@@ -0,0 +1,420 @@
+                let card: T = serde_json::from_slice(card_bytes.as_ref())?;
+                Ok(Some(card))
+            }
+            Ok(None) => Ok(None),
+            Err(err) => {
+                // TODO look at what errors NATS can give us and make more specific wrappers
+                Err(StorageError::NATSError(err.to_string()))
+            }
+        }
+    }
+
+    /// Returns a receiver that will receive all the existing keys, and
+    /// then block and receive new keys as they are created.
+    /// Starts a task that runs forever, watches the LiveStore.
+    ///
+    /// To get a KeyValueStore impl use either [`MemoryStorage`] for testing, or
+    /// nim_llm_pipeline::network::nats::kvstore::NATSStorage for production.
+    pub fn watch<T: for<'a> Deserialize<'a> + Send + 'static>(
+        self: Arc<Self>,
+        bucket_name: &str,
+        bucket_ttl: Option<Duration>,
+    ) -> (
+        tokio::task::JoinHandle<Result<(), StorageError>>,
+        tokio::sync::mpsc::UnboundedReceiver<T>,
+    ) {
+        let bucket_name = bucket_name.to_string();
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+        let watch_task = tokio::spawn(async move {
+            // Start listening for changes but don't poll this yet
+            let bucket = self
+                .0
+                .get_or_create_bucket(&bucket_name, bucket_ttl)
+                .await?;
+            let mut stream = bucket.watch().await?;
+
+            // Send all the existing keys
+            for (_, card_bytes) in bucket.entries().await? {
+                let card: T = serde_json::from_slice(card_bytes.as_ref())?;
+                let _ = tx.send(card);
+            }
+
+            // Now block waiting for new entries
+            while let Some(card_bytes) = stream.next().await {
+                let card: T = serde_json::from_slice(card_bytes.as_ref())?;
+                let _ = tx.send(card);
+            }
+
+            Ok::<(), StorageError>(())
+        });
+        (watch_task, rx)
+    }
+
+    pub async fn publish<T: Serialize + Versioned + Send + Sync>(
+        &self,
+        bucket_name: &str,
+        bucket_ttl: Option<Duration>,
+        key: &str,
+        obj: &mut T,
+    ) -> anyhow::Result<StorageOutcome> {
+        let obj_json = serde_json::to_string(obj)?;
+        let bucket = self.0.get_or_create_bucket(bucket_name, bucket_ttl).await?;
+
+        let outcome = bucket
+            .insert(key.to_string(), obj_json, obj.revision())
+            .await?;
+
+        match outcome {
+            StorageOutcome::Created(revision) | StorageOutcome::Exists(revision) => {
+                obj.set_revision(revision);
+            }
+        }
+        Ok(outcome)
+    }
+
+    /// Re-publish the model card to the store regularly. Spawns a task and returns.
+    /// Takes most arguments by value because it will hold on to them in the publish task.
+    /// Deletes the card on cancellation.
```

Comment:
What's the motivation or expected use case for regular re-publishing? If the object store or thing backing it (NATS in this specific case?) goes down?

Reply:
Good question. NATS persists to disk, and would be cluster in a production deployment, so we don't need to republish for that.

It's to handle the *worker* going away. We set a timeout on the model deployment card in NATS, at which point NATS deletes it. We re-publish before that happens. Hence if the worker dies NATS removes the card signaling the worker is unavailable.

The etcd concept of a "lease" is a much better match for what that's trying to achieve. Once this lands I will switch to etcd and remove the need to keep re-publishing. I did this work before we had etcd, and kept it to support NIM, but that's single node so they aren't publishing model deployment cards anyway.